include/ginkgo/core/base/device_matrix_data.hpp

```diff
@@ -234,7 +234,12 @@ class device_matrix_data {
      *
      * @param new_num_entries  the new number of stored matrix entries.
      */
-    void resize_and_reset(size_type new_num_entries);
+    void resize_and_reset(size_type new_num_entries)
+    {
+        row_idxs_.resize_and_reset(new_num_entries);
+        col_idxs_.resize_and_reset(new_num_entries);
+        values_.resize_and_reset(new_num_entries);
+    }
```

Comment:
So the issue is that I also need to set the dimension correctly, esp. for the offdiag. I think I can do that without calling the `resize_and_reset`, but it might be a bit cumbersome.

Reply:
The issue here is one of circular dependencies: core depends on device-specific libraries, so device-specific libraries can't call core functionality. This should be handled on the core level IMO. Pulling more functions into headers makes things difficult to untangle, also w.r.t. #715

Reply:
I generally agree with having a cleaner separation of .ccp/.hpp. Also my previous comment was wrong, and it's actually pretty straight forward to remove that call.